### PR TITLE
Added support for PHP7 \Throwable based exceptions

### PR DESCRIPTION
**Pull request by maurice2k**

### CODE REVIEW

File: src/DispatchListener.php [hunk 1]
```diff
@@ -109,15 +112,22 @@ public function onDispatch(MvcEvent $e)
 
         $request  = $e->getRequest();
         $response = $application->getResponse();
+        $caughtException = null;
 
         try {
             $return = $controller->dispatch($request, $response);
+        } catch (\Throwable $ex) {
```
Comment (manuakasam) [inline src/DispatchListener.php:119]: Shouldn't `Exception` be caught prior to `Throwable`? As Exception extends Throwable and would therefore never be caught?

Comment (maurice2k) [inline src/DispatchListener.php:119]: Makes no difference as the code is the same for `\Exception` and `\Throwable`. In PHP 7+ the exceptions or error exceptions are caught by `\Throwable` whereas in PHP 5 the exceptions are caught by `\Exception`.

`\Exception` should be removed at some point in the future (wenn PHP5 support is dropped).

Comment (Ocramius) [inline src/DispatchListener.php:119]: Can you add a `// @TODO clean up once PHP 7 requirement is enforced`?

File: src/MiddlewareListener.php [hunk 2]
```diff
@@ -56,9 +56,17 @@ public function onDispatch(MvcEvent $event)
             $event->setResult($return);
             return $return;
         }
+
+        $caughtException = null;
         try {
             $return = $middleware(Psr7Request::fromZend($request), Psr7Response::fromZend($response));
+        } catch (\Throwable $exception) {
```
Comment (Ocramius) [inline src/MiddlewareListener.php:63]: Same as above: TODO would be good

File: src/View/Console/ExceptionStrategy.php [hunk 3]
```diff
@@ -184,7 +184,7 @@ public function prepareExceptionViewModel(MvcEvent $e)
                 if (is_callable($this->message)) {
                     $callback = $this->message;
                     $message = (string) $callback($exception, $this->displayExceptions);
-                } elseif ($this->displayExceptions && $exception instanceof \Exception) {
+                } elseif ($this->displayExceptions && ($exception instanceof \Exception || $exception instanceof \Throwable)) {
```
Comment (Ocramius) [inline src/View/Console/ExceptionStrategy.php]: same

File: src/View/Console/RouteNotFoundStrategy.php [hunk 4]
```diff
@@ -460,7 +460,7 @@ protected function reportNotFoundReason($e)
         ];
         $report = sprintf("\nReason for failure: %s\n", $reasons[$reason]);
 
-        while ($exception instanceof \Exception) {
+        while ($exception instanceof \Exception || $exception instanceof \Throwable) {
```
Comment (Ocramius) [inline src/View/Console/RouteNotFoundStrategy.php]: same

File: src/View/Http/DefaultRenderingStrategy.php [hunk 5]
```diff
@@ -99,9 +99,17 @@ public function render(MvcEvent $e)
         $view->setRequest($request);
         $view->setResponse($response);
 
+        $caughtException = null;
+
         try {
             $view->render($viewModel);
+        } catch (\Throwable $ex) {
```
Comment (Ocramius) [inline src/View/Http/DefaultRenderingStrategy.php:106]: same

File: src/View/Http/RouteNotFoundStrategy.php [hunk 6]
```diff
@@ -250,7 +250,7 @@ protected function injectException($model, $e)
         $model->setVariable('display_exceptions', true);
 
         $exception = $e->getParam('exception', false);
-        if (!$exception instanceof \Exception) {
+        if (!$exception instanceof \Exception && !$exception instanceof \Throwable) {
```
Comment (Ocramius) [inline src/View/Http/RouteNotFoundStrategy.php]: same

File: src/DispatchListener.php (hunk 1 above)
Comment (maurice2k) [inline src/DispatchListener.php:119]: Good point, just added the todos.

File: src/View/Console/ExceptionStrategy.php [hunk 7]
```diff
@@ -184,7 +184,7 @@ public function prepareExceptionViewModel(MvcEvent $e)
                 if (is_callable($this->message)) {
                     $callback = $this->message;
                     $message = (string) $callback($exception, $this->displayExceptions);
-                } elseif ($this->displayExceptions && $exception instanceof \Exception) {
+                } elseif ($this->displayExceptions && ($exception instanceof \Exception || $exception instanceof \Throwable)) {  // @TODO clean up once PHP 7 requirement is enforced
```
Comment (weierophinney) [inline src/View/Console/ExceptionStrategy.php:187]: Can you open a related issue on [zend-mvc-console](https://github.com/zendframework/zend-mvc-console), please? This functionality is moved into there for the v3 series.

Comment (weierophinney) [inline src/View/Console/ExceptionStrategy.php:187]: Never mind; I'm working on it now.

**Pull request merged**